File: src/lib/sekiban-ts/src/client/executor.ts
import {
  type TagStateResponse,
  type CommandContext,
  type Command,
  tagString,
} from "../domain/types.js";

export class SekibanRuntimeClient {
  private baseUrl: string;
  private tagProjectorMap: Record<string, string>;

  constructor(baseUrl: string, tagProjectorMap: Record<string, string>) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.tagProjectorMap = tagProjectorMap;
  }

  private async postJSON(path: string, body: unknown): Promise<any> {
    const resp = await fetch(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (!resp.ok) {
      const errText = await resp.text();
      throw new Error(`${path} failed: ${resp.status} ${resp.statusText} - ${errText}`);
    }
    return await resp.json();
  }

  async getTagState(tagGroup: string, id: string): Promise<TagStateResponse> {
    const projectorName = this.tagProjectorMap[tagGroup];
    if (!projectorName) {
      throw new Error(`no projector mapped for tag group: ${tagGroup}`);
    }
    const tagStateId = `${tagGroup}:${id}:${projectorName}`;

    const resp: any = await this.postJSON("/api/sekiban/serialized/tag-state", {
      tagStateId,
    });

    let stateJson = "{}";
    if (resp.payload && resp.payload !== "") {
      try {
        const decoded = Buffer.from(resp.payload, "base64").toString("utf-8");
        if (decoded.length > 0) stateJson = decoded;
      } catch {
        // ignore decode errors
      }
    }

    const lastSortable = resp.lastSortableUniqueId || resp.lastSortedUniqueId || null;

    return {
      stateJson,
      version: resp.version ?? 0,
      lastSortableUniqueId: lastSortable,
    };
  }

  async executeQuery(
    queryType: string,
    paramsJson: string,
    serviceId?: string | null,
  ): Promise<string> {
    const body: Record<string, any> = {
      queryType,
      queryParamsJson: paramsJson,
    };
    if (serviceId) body.serviceId = serviceId;

    const resp = await this.postJSON("/api/sekiban/serialized/query", body);
    return resp.resultJson ?? "null";
  }

  async executeListQuery(
    queryType: string,
    paramsJson: string,
    serviceId?: string | null,
  ): Promise<string> {
    const body: Record<string, any> = {
      queryType,
      queryParamsJson: paramsJson,
    };
    if (serviceId) body.serviceId = serviceId;

    const resp = await this.postJSON("/api/sekiban/serialized/list-query", body);
    return resp.itemsJson ?? "[]";
  }

  async finalizeCommand(cmd: Command): Promise<any> {
    const cache = new Map<string, TagStateResponse>();

    const ctx: CommandContext = {
      cache,
      getTagState: async (
        tagGroup: string,
        id: string,
      ): Promise<TagStateResponse> => {
        const key = tagString(tagGroup, id);
        const cached = cache.get(key);
        if (cached) {
          return cached;
        }
        const state = await this.getTagState(tagGroup, id);
        cache.set(key, state);
        return state;
      },
    };

    const output = await cmd.handle(ctx);

    // Build event candidates
    const eventCandidates = output.events.map((ev) => {
      const payloadJson = JSON.stringify(ev.payload);
      const payloadBase64 = Buffer.from(payloadJson).toString("base64");
      return {
        eventPayloadName: ev.eventType,
        payload: payloadBase64,
        tags: ev.tags,
      };
    });

    // Build consistency tags from cached state
    const seen = new Set<string>();
    const consistencyTags: { tag: string; lastSortableUniqueId: string }[] = [];
    for (const cTag of output.consistencyTags) {
      if (seen.has(cTag)) continue;
      seen.add(cTag);
      const cached = cache.get(cTag);
      const lastId = cached?.lastSortableUniqueId ?? "";
      consistencyTags.push({
        tag: cTag,
        lastSortableUniqueId: lastId,
      });
    }

    const commitBody = {
      eventCandidates,
      consistencyTags,
    };

    return await this.postJSON("/api/sekiban/serialized/commit", commitBody);
  }
}

File: src/lib/sekiban-go/client/executor.go
package client

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/J-Tech-Japan/sekiban-go/domain"
)

const (
	defaultTimeout = 30 * time.Second
	maxRetries     = 3
)

// SekibanRuntimeClient is an HTTP client for the WasmServer serialized API.
type SekibanRuntimeClient struct {
	baseURL    string
	httpClient *http.Client
	// TagProjectorMap maps tag group -> projector name.
	TagProjectorMap map[string]string
}

// NewSekibanRuntimeClient creates a new client.
func NewSekibanRuntimeClient(baseURL string, tagProjectorMap map[string]string) *SekibanRuntimeClient {
	return &SekibanRuntimeClient{
		baseURL:         strings.TrimRight(baseURL, "/"),
		httpClient:      &http.Client{Timeout: defaultTimeout},
		TagProjectorMap: tagProjectorMap,
	}
}

// TagStateResult from the serialized tag-state endpoint.
type TagStateResult struct {
	Tag                  string
	TagStateId           string
	PayloadJson          string
	Version              int
	LastSortableUniqueId string
}

// GetTagState fetches tag state from the WasmServer.
func (c *SekibanRuntimeClient) GetTagState(tagGroup, tagId string) (TagStateResult, error) {
	projector := c.TagProjectorMap[tagGroup]
	if projector == "" {
		return TagStateResult{}, fmt.Errorf("unknown projector for tag group '%s'", tagGroup)
	}
	tagStateId := fmt.Sprintf("%s:%s:%s", tagGroup, tagId, projector)

	var resp struct {
		Payload              *string `json:"payload"`
		Version              int     `json:"version"`
		LastSortedUniqueId   string  `json:"lastSortedUniqueId"`
		LastSortableUniqueId string  `json:"lastSortableUniqueId"`
	}

	if err := c.postJSON("/api/sekiban/serialized/tag-state", map[string]string{"tagStateId": tagStateId}, &resp); err != nil {
		return TagStateResult{}, err
	}

	payloadJson := "{}"
	if resp.Payload != nil && *resp.Payload != "" {
		decoded, err := base64.StdEncoding.DecodeString(*resp.Payload)
		if err == nil && len(decoded) > 0 {
			payloadJson = string(decoded)
		}
	}

	lastSortable := resp.LastSortableUniqueId
	if lastSortable == "" {
		lastSortable = resp.LastSortedUniqueId
	}

	return TagStateResult{
		Tag:                  fmt.Sprintf("%s:%s", tagGroup, tagId),
		TagStateId:           tagStateId,
		PayloadJson:          payloadJson,
		Version:              resp.Version,
		LastSortableUniqueId: lastSortable,
	}, nil
}

// CommitRequest for the serialized commit endpoint.
type CommitRequest struct {
	Events []CommitEventCandidate `json:"events"`
}

// CommitEventCandidate is a single event in a commit request.
type CommitEventCandidate struct {
	EventPayloadName string   `json:"eventPayloadName"`
	Payload          string   `json:"payload"`
	Tags             []string `json:"tags"`
}

// ConsistencyTagEntry for the commit.
type ConsistencyTagEntry struct {
	Tag                  string `json:"tag"`
	LastSortableUniqueId string `json:"lastSortableUniqueId"`
}

// CommitResponse from the serialized commit endpoint.
type CommitResponse struct {
	EventId          *string `json:"eventId"`
	SortableUniqueId *string `json:"sortableUniqueId"`
}

// ExecuteListQuery executes a list query.
func (c *SekibanRuntimeClient) ExecuteListQuery(queryType, queryParamsJson string, waitForSortableUniqueId *string) (string, error) {
	body := map[string]any{
		"queryType":      queryType,
		"queryParamsJson": queryParamsJson,
	}
	if waitForSortableUniqueId != nil {
		body["waitForSortableUniqueId"] = *waitForSortableUniqueId
	}

	var resp struct {
		ItemsJson string `json:"itemsJson"`
	}
	if err := c.postJSON("/api/sekiban/serialized/list-query", body, &resp); err != nil {
		return "[]", err
	}
	return resp.ItemsJson, nil
}

// ExecuteQuery executes a scalar query.
func (c *SekibanRuntimeClient) ExecuteQuery(queryType, queryParamsJson string, waitForSortableUniqueId *string) (string, error) {
	body := map[string]any{
		"queryType":      queryType,
		"queryParamsJson": queryParamsJson,
	}
	if waitForSortableUniqueId != nil {
		body["waitForSortableUniqueId"] = *waitForSortableUniqueId
	}

	var resp struct {
		ResultJson string `json:"resultJson"`
	}
	if err := c.postJSON("/api/sekiban/serialized/query", body, &resp); err != nil {
		return "null", err
	}
	return resp.ResultJson, nil
}

// FinalizeCommand executes a command and commits the result.
func (c *SekibanRuntimeClient) FinalizeCommand(cmd domain.Command) (*CommitResponse, error) {
	ctx := &httpCommandContext{client: c}
	output, err := cmd.Handle(ctx)
	if err != nil {
		return nil, err
	}
	if output == nil || len(output.Events) == 0 {
		return &CommitResponse{}, nil
	}

	// Build event candidates
	eventCandidates := make([]CommitEventCandidate, 0, len(output.Events))
	for _, ev := range output.Events {
		encoded := base64.StdEncoding.EncodeToString([]byte(ev.Payload))
		eventCandidates = append(eventCandidates, CommitEventCandidate{
			EventPayloadName: ev.EventType,
			Payload:          encoded,
			Tags:             output.Tags,
		})
	}

	// Build consistency tags from loaded states
	seen := make(map[string]bool)
	consistencyTags := make([]ConsistencyTagEntry, 0, len(output.ConsistencyTags))
	for _, tag := range output.ConsistencyTags {
		if seen[tag] {
			continue
		}
		seen[tag] = true
		lastSortable := ""
		if cached, ok := ctx.cache[tag]; ok && cached.LastSortableUniqueId != nil {
			lastSortable = *cached.LastSortableUniqueId
		}
		consistencyTags = append(consistencyTags, ConsistencyTagEntry{
			Tag:                  tag,
			LastSortableUniqueId: lastSortable,
		})
	}

	commitBody := map[string]any{
		"eventCandidates": eventCandidates,
		"consistencyTags": consistencyTags,
	}

	var commitResp CommitResponse
	if err := c.postJSON("/api/sekiban/serialized/commit", commitBody, &commitResp); err != nil {
		return nil, err
	}
	return &commitResp, nil
}

// httpCommandContext implements CommandContext via HTTP.
type httpCommandContext struct {
	client *SekibanRuntimeClient
	cache  map[string]domain.TagStateResponse
}

func (ctx *httpCommandContext) GetTagState(tagGroup, tagId string) (domain.TagStateResponse, error) {
	cacheKey := tagGroup + ":" + tagId
	if ctx.cache != nil {
		if cached, ok := ctx.cache[cacheKey]; ok {
			return cached, nil
		}
	}

	result, err := ctx.client.GetTagState(tagGroup, tagId)
	if err != nil {
		return domain.TagStateResponse{}, err
	}

	resp := domain.TagStateResponse{
		TagGroup:  tagGroup,
		TagId:     tagId,
		StateJson: result.PayloadJson,
		Version:   result.Version,
	}
	if result.LastSortableUniqueId != "" {
		resp.LastSortableUniqueId = &result.LastSortableUniqueId
	}

	if ctx.cache == nil {
		ctx.cache = make(map[string]domain.TagStateResponse)
	}
	ctx.cache[cacheKey] = resp
	return resp, nil
}

// HTTP helpers

func (c *SekibanRuntimeClient) postJSON(path string, payload any, target any) error {
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	var lastErr error
	for attempt := 0; attempt < maxRetries; attempt++ {
		resp, err := c.httpClient.Post(c.baseURL+path, "application/json", bytes.NewReader(data))
		if err != nil {
			lastErr = err
			continue
		}
		defer resp.Body.Close()
		responseBytes, _ := io.ReadAll(resp.Body)
		if resp.StatusCode >= 400 {
			return fmt.Errorf("http %d: %s", resp.StatusCode, string(responseBytes))
		}
		if target == nil {
			return nil
		}
		return json.Unmarshal(responseBytes, target)
	}
	if lastErr != nil {
		return lastErr
	}
	return errors.New("request failed after retries")
}
